src/Domain/Services/ExtendRestApi.php

```diff
@@ -118,6 +126,66 @@ public function register_endpoint_data( $args ) {
 		return true;
 	}
 
+	/**
+	 * Add callback functions that can be executed by the cart/extensions endpoint.
+	 *
+	 * @param array $args {
+	 *     An array of elements that make up the callback configuration.
+	 *
+	 *     @type string   $endpoint The endpoint to extend.
+	 *     @type string   $namespace Plugin namespace.
+	 *     @type callable $callback_function The function/callable to execute.
+	 * }
+	 *
+	 * @throws Exception On failure to register.
+	 * @returns boolean True on success.
+	 */
+	public function register_update_callback( $args ) {
+		if ( ! is_string( $args['namespace'] ) ) {
+			$this->throw_exception( 'You must provide a plugin namespace when extending a Store REST endpoint.' );
+		}
+
+		if ( ! is_string( $args['endpoint'] ) || ! in_array( $args['endpoint'], $this->endpoints, true ) ) {
+			$this->throw_exception(
+				sprintf( 'You must provide a valid Store REST endpoint to extend, valid endpoints are: %1$s. You provided %2$s.', implode( ', ', $this->endpoints ), $args['endpoint'] )
+			);
+		}
+
+		$this->callback_methods[ $args['endpoint'] ][ $args['namespace'] ] = [
+			'callback_function' => $args['callback_function'],
+		];
+		return true;
+	}
+
+	/**
+	 * Get callback for a specific endpoint and namespace.
+	 *
+	 * @param string $endpoint The endpoint to get callbacks for.
+	 * @param string $namespace The namespace to get callbacks for.
+	 *
+	 * @return callable The callback registered by the extension.
+	 * @throws Exception When callback is not callable or parameters are incorrect.
+	 */
+	public function get_update_callback( $endpoint, $namespace ) {
+		if ( ! is_string( $namespace ) ) {
+			$this->throw_exception( 'You must provide a plugin namespace when extending a Store REST endpoint.' );
+		}
+
+		if ( ! is_string( $endpoint ) || ! in_array( $endpoint, $this->endpoints, true ) ) {
+			$this->throw_exception(
+				sprintf( 'You must provide a valid Store REST endpoint to extend, valid endpoints are: %1$s. You provided %2$s.', implode( ', ', $this->endpoints ), $endpoint )
+			);
+		}
+
+		$method = $this->callback_methods[ $endpoint ][ $namespace ]['callback_function'];
+		if ( ! is_callable( $method ) ) {
+			$this->throw_exception(
+				sprintf( 'There is no valid callback registered for: %1$s on the %2$s endpoint.', $namespace, $endpoint )
+			);
+		}
+		return $method;
+	}
```

Comment:
Who executes `$method` here? I think it should be executed at `ExtendRestApi` level and for it to wrapped in a try/catch. We should return an Error Response to users. a Try Catch would just make sure the registration doesn't fail breaking the whole API.

Reply:
`$method` is executed by `get_item_response` in `CartExtensionSchema` I have added a try catch there in 468bae3d20a46ebf9334f4f62d36809e0cc64941. This would cause the promise returned by `extensionCartUpdate` to reject. Extensions can then catch this and handle it.

![image](https://user-images.githubusercontent.com/5656702/120517474-662c8000-c3c8-11eb-82f0-9a91674b627c.png)

Is this what you were thinking of or did you have something else in mind?